File: netCDF/examples/F90/example_par_write.f90
program example_par_write
  use netcdf
  implicit none
  include 'mpif.h'

  integer :: mpirank
  integer :: mpisize
  integer :: ierr

  character (len = *), parameter :: FILE_NAME = "par_write.nc"
  integer,parameter :: NDIMS=2
  integer,parameter :: NI_GLO=5
  integer,parameter :: NJ_GLO=10

  integer :: field_glo(NI_GLO,NJ_GLO)
  integer,allocatable :: field(:,:)
  integer :: ni,ibegin,nj,jbegin
  integer :: ncid,idimid,jdimid,varid

  !--------------------------------------
  ! MPI initialization
  !--------------------------------------

  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,mpirank,ierr)
  call MPI_Comm_size(MPI_COMM_WORLD,mpisize,ierr)

  !--------------------------------------
  !  NetCDF metadata definitions
  !--------------------------------------

  ! Create a file
  ierr = nf90_create(FILE_NAME, IOR(NF90_NETCDF4,NF90_MPIIO), ncid,&
                    comm=MPI_COMM_WORLD, info=MPI_INFO_NULL)

  ! Define two spatial dimensions
  ierr = nf90_def_dim(ncid, "ni", NI_GLO, idimid)
  ierr = nf90_def_dim(ncid, "nj", NJ_GLO, jdimid)

  ! Define a 2D variable and add an attribute
  ierr = nf90_def_var(ncid, "toto", NF90_INT, (/idimid, jdimid/), varid)
  ierr = nf90_put_att(ncid, varid, "standart_name", "Dummy variable")

  ! Close metadata definition
  ierr = nf90_enddef(ncid)

  !--------------------------------------
  !  Global and local data initialization
  !--------------------------------------

  ! Initialize global data array
  !
  call init_global(field_glo)

  ! Decompose global domain.
  ! Returned values: ni,nj,ibegin,jbegin.
  !
  call decompose_domain(mpisize,mpirank,ni,ibegin,nj,jbegin)

  ! Allocate and fill in local array
  !
  allocate(field(ni,nj))
  field(:,:) = field_glo(ibegin:ibegin-1+ni,jbegin:jbegin-1+nj)
  
  !--------------------------------------
  ! Write data and close the file  
  !--------------------------------------

  ! Write data
  ierr = nf90_put_var(ncid, varid, field, start=(/ibegin,jbegin/), count=(/ni,nj/))
  ! Close the file
  ierr = nf90_close(ncid)
 
  deallocate(field)

  call MPI_FINALIZE(ierr)

  !--------------------------------------
  ! Functions and subroutines
  !--------------------------------------

  contains 

  !--------------------------------------
  ! This subroutine fills in global data array
  !--------------------------------------

  subroutine init_global(field_glo)
  integer,intent(inout) :: field_glo(NI_GLO,NJ_GLO)
  integer i,j

  do j=1,NJ_GLO
    do i=1,NI_GLO
      field_glo(i,j)=(i-1)+(j-1)*NI_GLO
    enddo
  enddo
  end subroutine init_global

  !--------------------------------------
  ! This subroutine decomposes global domain
  !--------------------------------------
  subroutine decompose_domain(mpisize,rank,ni,ibegin,nj,jbegin)
    integer,intent(in) :: mpisize,rank
    integer,intent(out) :: ni,ibegin,nj,jbegin
    integer :: n
    ni=NI_GLO ; ibegin=1
    jbegin=0
    do n=0,mpisize-1
      nj=NJ_GLO/mpisize
      if (n<MOD(NJ_GLO,mpisize)) nj=nj+1
      if (n==rank) exit
      jbegin=jbegin+nj
    enddo
    jbegin=jbegin+1
  end subroutine decompose_domain

end program example_par_write

File: netCDF/examples/C/example_par_write.c
#include "mpi.h"
#include <netcdf.h>
#include <netcdf_par.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NI_GLO    5 
#define NJ_GLO    10
#define NDIMS 2

#define ERR(e) {printf("Error: %s\n", nc_strerror(e)); return 2;}


// Function decomposes global domain of size NI_GLO*NJ_GLO in horizontal direction into local domains
void decompose_domain(int mpisize, int mpirank, int* ni, int* ibegin, int* nj, int* jbegin)
{
    *nj=NJ_GLO;
    *ibegin=0;
    *jbegin=0;
    for (int n=0; n<mpisize; ++n)
    {     
      *ni=NI_GLO/mpisize;
      if (n<(NI_GLO%mpisize)) ++(*ni);
      if (n==mpirank) break;
      (*ibegin)+=(*ni);
    }
}

// Funciton fills in local domain data
void init_local_data(int* data, int ni, int ibegin, int nj, int jbegin)
{
  for(int i=0; i < ni; i++)
    for(int j=0; j < nj; j++)
      data[j+i*nj] = (j+jbegin) + (i+ibegin)*nj;
}

int main (void)
{
  int dimids[NDIMS];
  size_t starts[NDIMS], counts[NDIMS];
  int mpirank,mpisize;
  int ncid,xdimid,ydimid,varid;
  int err;
  int ni, nj, ibegin, jbegin;
  char filename[128] = "par_write.nc";

  //--------------------------------------
  // Initialization MPI
  //--------------------------------------

  MPI_Init(NULL, NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&mpirank);
  MPI_Comm_size(MPI_COMM_WORLD,&mpisize);

  //--------------------------------------
  //  NetCDF metadata definitions
  //--------------------------------------

  // Create a netcdf file
  if (err=nc_create_par(filename, NC_NETCDF4, MPI_COMM_WORLD, MPI_INFO_NULL, &ncid)) ERR(err);

  // Define two spatial dimensions
  if (err=nc_def_dim(ncid, "ni", NI_GLO, &xdimid)) ERR(err);
  if (err=nc_def_dim(ncid, "nj", NJ_GLO, &ydimid)) ERR(err);
  
  // Define a 2D variable and add an attribute
  dimids[0] = xdimid; 
  dimids[1] = ydimid;
  if (err=nc_def_var(ncid, "toto", NC_INT, NDIMS, dimids, &varid)) ERR(err);
  if (err=nc_put_att(ncid, varid, "standart_name", NC_CHAR, strlen("Dummy variable"), "Dummy variable")) ERR(err);

  // Close metadata definition
  nc_enddef(ncid);

  //--------------------------------------
  // Local data initialization
  //--------------------------------------

  // Decompose global domain
  // Returned values: ni,nj,ibegin,jbegin
  decompose_domain(mpisize,mpirank,&ni,&ibegin,&nj,&jbegin);
  
  // Initialize local data
  int *data = (int *)malloc(sizeof(int)*ni*nj);
  init_local_data(data, ni, ibegin, nj, jbegin);

  //--------------------------------------
  // Write data and close the file  
  //--------------------------------------

  // Write data
  starts[0] = ibegin; starts[1] = jbegin;
  counts[0] = ni;     counts[1] = nj;
  if (err=nc_put_vara(ncid, varid, starts, counts, data)) ERR(err);
  
  // Close the file
  if (err=nc_close(ncid)) ERR(err);

  free(data);

  MPI_Finalize();

  return 0;
}     

